### PR TITLE
Add default `confidence` param to sssom file while creating `ptable`

### PR DESCRIPTION
**Pull request by hrshdhgd**
If the confidence column is absent, `ptable` cannot be generated.

Fix: Add the `confidence` column if absent and initialize it to a user provided value. If present, then fill `None`s by that value

### CODE REVIEW

File: src/sssom/cli.py [hunk 1]
```diff
@@ -253,14 +254,25 @@ def split(input: str, output_directory: str):
 @input_argument
 @output_option
 @click.option("-W", "--inverse-factor", help="Inverse factor.")
-def ptable(input, output: TextIO, inverse_factor):
+@click.option(
+    "--default-confidence / --no-default-confidence",
+    default=False,
+    is_flag=True,
```
Comment (matentzn) [inline src/sssom/cli.py:260]: 1. Dont make it a flag, just make it 

```
--confidence X
```

where X is a number, e.g.

```
--confidence 0.9
```

Much more flexible, especially if you deal with mapping sets you trust differently!

Also, I thought about it and we can catch two flies with one here: Lets say this is the confidence of the person processing the mapping. We can use this not only to fill in missing confidence, but perhaps (in a second step, not this PR) to re-calculate existing confidence values.

So lets say we have a mapping that already has 0.8 confidence. If the mapping user provides `--confidence 0.9` then we could recalcularte the confidence like this `0.8*0.9=0.71` I think that is useful for many use cases..

File: src/sssom/util.py [hunk 2]
```diff
@@ -504,17 +504,38 @@ def compare_dataframes(df1: pd.DataFrame, df2: pd.DataFrame) -> MappingSetDiff:
...
+    :param df: DataFrame whose `confidence` column needs to be filled.
+    :return: DataFrame with a complete `confidence` column.
+    """
+    df[CONFIDENCE] = df.get(CONFIDENCE, 0.95)
+    df.loc[df[CONFIDENCE].isnull(), CONFIDENCE] = 0.95
+    return df
+
+
+def dataframe_to_ptable(
+    df: pd.DataFrame, *, inverse_factor: float = None, default_confidence: bool = False
```
Comment (matentzn) [inline src/sssom/util.py:521]: Rename to just "confidence"?

Comment (hrshdhgd) [reply src/sssom/util.py:521]: I'll rename to `default_confidence` since there's already a variable named `confidence` down below in the same function.

File: src/sssom/cli.py (hunk 1 above)
Comment (hrshdhgd) [reply src/sssom/cli.py:260]: So in other words if an external confidence is provided (say 0.9 in your example) and existing confidence reduces (0.72)?

Comment (hrshdhgd) [reply src/sssom/cli.py:260]: I can do recalculation in this PR itself, it's very simple!

File: src/sssom/cli.py [hunk 3]
```diff
@@ -253,14 +254,23 @@ def split(input: str, output_directory: str):
...
+    default=None,
+    help="Default confidence to be assigned if absent.",
+)
+def ptable(input, output: TextIO, inverse_factor: float, default_confidence: float):
     """Convert an SSSOM file to a ptable for kboom/`boomer <https://github.com/INCATools/boomer>`_."""
     # TODO should maybe move to boomer (but for now it can live here, so cjm can tweak
     msdf = parse_sssom_table(input)
     # df = parse(input)
+    if default_confidence:
+        msdf.df = add_default_confidence(msdf.df, default_confidence)
```
Comment (matentzn) [inline src/sssom/cli.py:268]: Why is this added twice? Once here, and once in `dataframe_to_ptable`?

File: src/sssom/util.py [hunk 4]
```diff
@@ -504,17 +504,39 @@ def compare_dataframes(df1: pd.DataFrame, df2: pd.DataFrame) -> MappingSetDiff:
...
+
+    If `confidence` column already exists, only fill in the None ones by 0.95.
+
+    :param df: DataFrame whose `confidence` column needs to be filled.
+    :return: DataFrame with a complete `confidence` column.
+    """
+    # df[CONFIDENCE] = df.get(CONFIDENCE, confidence)
+    if df.get(CONFIDENCE) is not None:
+        df[CONFIDENCE] = confidence * df[CONFIDENCE]
+    df.loc[df[CONFIDENCE].isnull(), CONFIDENCE] = confidence
```
Comment (matentzn) [inline src/sssom/util.py:518]: Is this doing the exact same thing as this (which is how I thought of the problem):

```
   if 'confidence' not in df.columns:
        df['confidence'] = confidence
    else:
        df['confidence'] = np.where(
            (df['confidence'].between(0, 1)) & (pd.notnull(df['confidence'])),
            df['confidence'] * confidence,
            confidence
        )
```

File: src/sssom/cli.py (hunk 3 above)
Comment (hrshdhgd) [reply src/sssom/cli.py:268]: Yeah I had that question too... this is how it was from the beginning. Let me streamline it.

Comment (hrshdhgd) [reply src/sssom/cli.py]: Done

File: src/sssom/util.py (hunk 4 above)
Comment (hrshdhgd) [reply src/sssom/util.py:518]: Yes

File: src/sssom/cli.py [hunk 5]
```diff
@@ -253,14 +252,18 @@ def split(input: str, output_directory: str):
...
+    "--default-confidence",
+    default=None,
+    help="Default confidence to be assigned if absent.",
+)
+def ptable(input, output: TextIO, inverse_factor: float, default_confidence: float):
     """Convert an SSSOM file to a ptable for kboom/`boomer <https://github.com/INCATools/boomer>`_."""
     # TODO should maybe move to boomer (but for now it can live here, so cjm can tweak
     msdf = parse_sssom_table(input)
-    # df = parse(input)
-    df = collapse(msdf.df)
```
Comment (matentzn) [inline src/sssom/cli.py:261]: ? Why is collapse removed?

Comment (matentzn) [reply src/sssom/cli.py:261]: Because this already happens in `dataframe_to_ptable`.

File: src/sssom/cli.py (hunk 3 above)
Comment (matentzn) [reply src/sssom/cli.py]: Great.

File: src/sssom/cli.py [hunk 6]
```diff
@@ -253,14 +252,18 @@ def split(input: str, output_directory: str):
 @input_argument
 @output_option
 @click.option("-W", "--inverse-factor", help="Inverse factor.")
-def ptable(input, output: TextIO, inverse_factor):
+@click.option(
+    "--default-confidence",
+    default=None,
```
Comment (matentzn) [inline src/sssom/cli.py:257]:
```suggestion
    default=None,
    type=click.FloatRange(0, 1),
```

Not sure if that works, maybe check

File: src/sssom/util.py [hunk 7]
```diff
@@ -504,17 +504,40 @@ def compare_dataframes(df1: pd.DataFrame, df2: pd.DataFrame) -> MappingSetDiff:
...
+def add_default_confidence(df: pd.DataFrame, confidence: float = None) -> pd.DataFrame:
+    """Add `confidence` column to DataFrame if absent and initializes to 0.95.
+
+    If `confidence` column already exists, only fill in the None ones by 0.95.
+
+    :param df: DataFrame whose `confidence` column needs to be filled.
+    :return: DataFrame with a complete `confidence` column.
+    """
+    if CONFIDENCE in df.columns:
+        df[CONFIDENCE] = confidence * df.get(CONFIDENCE, 1)
```
Comment (matentzn) [inline src/sssom/util.py:516]: What happens if there is a 

NaN (a non numeric string)
None or ""
Or NULL

In the `confidence` column?

File: src/sssom/cli.py (hunk 5 above)
Comment (hrshdhgd) [reply src/sssom/cli.py:261]: Exactly

File: src/sssom/cli.py (hunk 6 above)
Comment (hrshdhgd) [reply src/sssom/cli.py:257]: I've changed it as per click documentation

File: src/sssom/util.py (hunk 7 above)
Comment (hrshdhgd) [reply src/sssom/util.py]: Good catch! Fixed it

**Review by matentzn — DISMISSED** [review]
Fantastic effort! THANK YOU!
